text/0779-first-class-component-templates.md

```diff
@@ -0,0 +1,1622 @@
+
+```js
+import GreetingComponent from './greeting';
+
+<template>
+  <GreetingComponent @name={{@user.name}} />
+</template>
+```
+
+This is obviously undesirable, but avoiding this will mean mean renaming locally after the auto-complete works. That renaming operation is a needless paper cut in the best case of importing a default export. It rises to the level of a significant annoyance when using named imports:
+
+```js
+import {
+  ButtonComponent as UIButton,
+  FormComponent as UIForm,
+  InputComponent as UIInput,
+} from './ui';
+
+<template>
+  <UIForm @onSubmit={{@saveName}}>
+    <UIInput @label="Name" @value={{@name}} @onUpdate={{@updateName}} />
+    <UIButton @label="Save" type="submit" />
+  </UIForm>
+</template>
+```
+
+And it makes namespace-style imports basically unusable: to invoke *without* `Component` everywhere, you have to rebind all the imports you use!
+
+```js
+import * as _UI from './ui';
+const UI = {
+  Form: _UI.FormComponent,
+  Button: _UI.ButtonComponent,
+  Input: _UI.InputComponent
+}
+
+<template>
+  <UI.Form @onSubmit={{@saveName}}>
+    <UI.Input @label="Name" @value={{@name}} @onUpdate={{@updateName}} />
+    <UI.Button @label="Save" type="submit" />
+  </UI.Form>
+</template>
+```
+
+Accordingly, we should switch to generating *without* a class name: `ember generate component greeting --component-class=@glimmer/component` should produce a class named `Greeting`, *not* `GreetingComponent`. The generated names for routes, services, and controllers can remain as they are, since they are never invoked this way.
+  
+[^polaris]: Polaris was announced as planned at EmberConf 2021. This plan assumes we ship Polaris before Ember 5. If we ship Ember 5 first, the dynamics would be much the same, but with the major version as the point when we switch the default instead.
+
+[^ts-component-name]: In TypeScript, this also extends to `GreetingComponentArgs` (or, with [RFC #0748][rfc-0748], something like `GreetingComponentSignature`), which gets *really* unwieldy!
+
+
+#### Linting and formatting
+
+As with syntax highlighting, we need to support the new format with linting and formatting integration.
+
+For linting, we need to make two changes:
+
+1. Create an ESLint processor which uses the same Babel transform as the core behavior itself (as provided currently by [ember-template-imports][eti]) to make normal ESLint rules (e.g. unused imports and values) work with the template scope.
```

Comment:
I'm a bit concerned that this means that things like line numbers _below_ the `<template>` tag will be "wrong" for other warnings/errors that happen.

Reply:
This is a real possibility, but the tooling affords us the ability to supply offsets when doing the transform. We will need to make sure that we handle that across the board: it's already something Glint has to do, and ESLint and similar will have to accommodate it as well, so we should try to share the relevant transform infrastructure so that they can all take advantage of it.